### PR TITLE
[CARBONDATA-2819] Fixed cannot drop preagg datamap on table if the table has other type datamaps

### PR DESCRIPTION
**Pull request by Sssan520**
1.Use "SET carbon.datamap.visible.{dbName}.{mainTable}.{datamapName} = {true or false}",if datamap provider is 'preagg',block this configuration.
2.When create preagg datamap, now also create its datamap schema file in system folder.
3.Use command "show datamap on table {tableName}" will get its all datamaps from datamap schema files

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?No
 
 - [ ] Any backward compatibility impacted? No
 
 - [ ] Document update required? No

 - [ ] Testing done 
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7833/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6558/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6207/

**Comment by Sssan520** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7839/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6563/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7841/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6214/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6565/

**Comment by Sssan520** [comment]
retest this please

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6215/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7842/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6567/

**Comment by Sssan520** [comment]
retest this please

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6218/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7847/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6572/

**Comment by Sssan520** [comment]
retest this please

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6219/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7848/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6573/

File: core/src/main/java/org/apache/carbondata/core/util/SessionParams.java [hunk 1]
```diff
@@ -213,13 +217,20 @@ private boolean validateKeyValue(String key, String value) throws InvalidConfigu
...
           String[] keyArray = key.split("\\.");
-          isValid = DataMapStoreManager.getInstance().isDataMapExist(
-              keyArray[keyArray.length - 3],
-              keyArray[keyArray.length - 2],
-              keyArray[keyArray.length - 1]);
-          if (!isValid) {
+          try {
+            DataMapSchema dataMapSchema =
+                DataMapStoreManager.getInstance().getDataMapSchema(keyArray[keyArray.length - 1]);
+            if (DataMapClassProvider.PREAGGREGATE.getShortName()
```
Comment (xuchuanyin) [inline core/src/main/java/org/apache/carbondata/core/util/SessionParams.java]: This command is only works for index datamap, so please optimize it again

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/PreAggregateDataMapExample.scala [hunk 2]
```diff
@@ -175,7 +175,7 @@ object PreAggregateDataMapExample {
 
     // create pre-aggregate table by datamap
     spark.sql("""
-       CREATE datamap preagg_avg on table personTable using 'preaggregate' as
+       CREATE datamap preagg_avg_dm on table personTable using 'preaggregate' as
```
Comment (xuchuanyin) [inline examples/spark2/src/main/scala/org/apache/carbondata/examples/PreAggregateDataMapExample.scala]: why this modification is needed?

File: integration/spark2/src/main/java/org/apache/carbondata/datamap/PreAggregateDataMapProvider.java [hunk 3]
```diff
@@ -77,13 +79,15 @@ private void validateDmProperty(DataMapSchema dataMapSchema)
...
-  public void cleanMeta() {
+  public void cleanMeta() throws IOException {
+    DataMapSchema dataMapSchema = getDataMapSchema();
     dropTableCommand = new CarbonDropTableCommand(
         true,
         new Some<>(dbName),
         tableName,
         true);
     dropTableCommand.processMetadata(sparkSession);
+    DataMapStoreManager.getInstance().dropDataMapSchema(dataMapSchema.getDataMapName());
```
Comment (xuchuanyin) [inline integration/spark2/src/main/java/org/apache/carbondata/datamap/PreAggregateDataMapProvider.java]: Can you check how MV deal with this? In another word, what's the logic of dropping a MV?

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7851/

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/CarbonDropDataMapCommand.scala [hunk 4]
```diff
@@ -114,12 +114,6 @@ case class CarbonDropDataMapCommand(
         LOGGER.audit(s"Deleting datamap [$dataMapName] under table [$tableName]")
 
         // drop index datamap on the main table
-        if (mainTable != null &&
```
Comment (xuchuanyin) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/CarbonDropDataMapCommand.scala]: Can you explain why this modification is needed?

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6577/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6605/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7882/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6244/

File: examples/spark2/src/main/scala/org/apache/carbondata/examples/PreAggregateDataMapExample.scala (hunk 2 above)
Comment (Sssan520) [reply examples/spark2/src/main/scala/org/apache/carbondata/examples/PreAggregateDataMapExample.scala]: As datamap name is a global, there is already the same  datamap name before.

File: integration/spark2/src/main/java/org/apache/carbondata/datamap/PreAggregateDataMapProvider.java (hunk 3 above)
Comment (Sssan520) [reply integration/spark2/src/main/java/org/apache/carbondata/datamap/PreAggregateDataMapProvider.java]: when create a mv datamap, will create a datamap scheme file and register a datamap catalog; then when drop the mv datamap,will delete the datamap schema file and unregister the datamap catalog.

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/CarbonDropDataMapCommand.scala (hunk 4 above)
Comment (Sssan520) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/CarbonDropDataMapCommand.scala]: when drop preagg datamap  , should delete the datamap schema file and main tableinfo schema info.

**Comment by Sssan520** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7883/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6607/

**Comment by Sssan520** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7901/

**Pull request closed without merging**